multinsclient/client.go

```diff
@@ -0,0 +1,177 @@
+	return int(h.Sum32()) % max
+}
+
+// New creates a new client from the given options.
+func New(options Options) (*Client, error) {
+	c := &Client{
+		hasher:                  options.Hasher,
+		enabledNamespaceClients: map[string]client.Client{},
+	}
+
+	// Use default hasher if none given
+	if c.hasher == nil {
+		c.hasher = DefaultHasher
+	}
+
+	// Close any created clients of this didn't complete successfully
+	success := false
+	defer func() {
+		if !success {
+			for _, client := range c.enabledNamespaceClients {
+				client.Close()
+			}
+		}
+	}()
+
+	// Create all non-disabled clients
+	seenNamespaces := make(map[string]bool, len(options.Namespaces))
+	for _, namespace := range options.Namespaces {
+		// We don't support default namespace name fallback, we require it to be set
+		if namespace.ClientOptions.Namespace == "" {
+			return nil, fmt.Errorf("all namespaces must have a namespace in the client options")
+		}
+		if _, exists := seenNamespaces[namespace.ClientOptions.Namespace]; exists {
+			return nil, fmt.Errorf("duplicate namespace %q", namespace.ClientOptions.Namespace)
+		}
+		seenNamespaces[namespace.ClientOptions.Namespace] = true
+		c.allNamespaces = append(c.allNamespaces, namespace.ClientOptions.Namespace)
+		if !namespace.Disabled {
+			c.enabledNamespaces = append(c.enabledNamespaces, namespace.ClientOptions.Namespace)
+			var err error
+			// TODO(cretz): Could alter options to have client options unrelated to
+			// namespaces and then use the new client.NewClientFromExisting introduced
+			// in https://github.com/temporalio/sdk-go/pull/881 to share connections
+			// across namespaces, but having multiple clients is mostly harmless.
+			if options.ConnectLazy {
+				c.enabledNamespaceClients[namespace.ClientOptions.Namespace], err = client.NewLazyClient(namespace.ClientOptions)
+			} else {
+				c.enabledNamespaceClients[namespace.ClientOptions.Namespace], err = client.Dial(namespace.ClientOptions)
+			}
+			if err != nil {
+				return nil, fmt.Errorf("failed creating client for namespace %q: %w", namespace.ClientOptions.Namespace, err)
+			}
+		}
+	}
+
+	// Must be at least one enabled
+	if len(c.enabledNamespaces) == 0 {
+		return nil, fmt.Errorf("no enabled namespaces")
+	}
+
+	success = true
+	return c, nil
+}
+
+// NamespaceFor selects a namespace and client for the given workflow ID. The
+// resulting namespace is always the same for the same workflow ID.
+//
+// If onlyEnabled is true and the namespace that would be selected is not
+// enabled, a hash is done only across the enabled namespace. Therefore, while
+// the same namespace is always selected even after Reset for the same workflow
+// ID when onlyEnabled is false, when onlyEnabled is true, the selected
+// namespace may be different for each Reset.
+//
+// The resulting client is only non-nil if the namespace is enabled.
+func (c *Client) NamespaceFor(workflowID string, onlyEnabled bool) (namespace string, client client.Client) {
```

Comment:
Do we actually need `onlyEnabled`? If there's no scenario where we expect it to be passed as `false` , we can simplify by taking it out and always falling back to enabled in case the primary match is not.

Reply:
If y'all are too concerned about having this available, I can make this method unexported. It's not like this is some complicated library I need to hide things in. If I were using this library, I might to know which namespace _would_ have been picked if all were enabled.

Reply:
Yes, make this method unexpected, and have another one that exported that call into this one with the flag always set to true.

Reply:
Why? Why can't I as a user get what the namespace would be if they were all enabled? There is no benefit in limiting the user here.